File: src/same-tree/same-tree.c
#include "../tree_node.h"
#include "stdbool.h"
#include "assert.h"

bool isSameTree(struct TreeNode* a, struct TreeNode* b) {
    if (a == NULL && b == NULL) {
        return true;
    }

    if ((a == NULL && b != NULL)
        || (b == NULL && a != NULL)
        || (a->val != b->val)
    ) {
        return false;
    }

    return isSameTree(a->left, b->left)
        && isSameTree(a->right, b->right);
}


int main(void) {
    struct TreeNode a0, a1, a2;
    init_tree_node(&a0, 0, &a1, &a2);
    init_tree_node(&a1, 1, NULL, NULL);
    init_tree_node(&a2, 2, NULL, NULL);
    
    struct TreeNode b0, b1, b2;
    init_tree_node(&b0, 0, &b1, &b2);
    init_tree_node(&b1, 1, NULL, NULL);
    init_tree_node(&b2, 2, NULL, NULL);

    bool res = isSameTree(&a0, &b0);
    printf("res = %d\n", res);
    assert(res);

    return 0;
}

File: n150/same-tree/same-tree.py
# https://leetcode.com/problems/same-tree

from TreeNode import TreeNode

class Solution(object):
    def step(self, p, q):
        if p == None and q == None:
            return True
        if p == None or q == None:
            return False
        if p.val != q.val:
            return False
        a = self.step(p.left, q.left)
        b = self.step(p.right, q.right)
        return a and b
        
    def isSameTree(self, p, q):
        return self.step(p, q)

if __name__ == "__main__":
    sol = Solution()
    tests = [
        (
            [1,2,3],
            [1,2,3],
            True,
        ),
        (
            [1,2],
            [1,None,2],
            False,
        ),
        (
            [1,2,1],
            [1,1,2],
            False,
        ),
    ]
    for (p, q, solution) in tests:
        print("----------------------")
        p = TreeNode.from_array(p)
        print("p:")
        p.print_tree()
        q = TreeNode.from_array(q)
        print("q:")
        q.print_tree()
        result = sol.isSameTree(p, q)
        assert result == solution, \
            f"result {result} != solution {solution}"
    print("✅ All tests passed")

